File: src/main/java/pl/allegro/experiments/chi/chiserver/domain/experiments/administration/start/StartExperimentCommandFactory.java
package pl.allegro.experiments.chi.chiserver.domain.experiments.administration.start;

import com.google.common.base.Preconditions;
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentsRepository;
import pl.allegro.experiments.chi.chiserver.domain.experiments.administration.PermissionsAwareExperimentRepository;

public class StartExperimentCommandFactory {
    private final ExperimentsRepository experimentsRepository;
    private final PermissionsAwareExperimentRepository permissionsAwareExperimentRepository;

    public StartExperimentCommandFactory(
            ExperimentsRepository experimentsRepository,
            PermissionsAwareExperimentRepository permissionsAwareExperimentRepository) {
        Preconditions.checkNotNull(experimentsRepository);
        Preconditions.checkNotNull(permissionsAwareExperimentRepository);
        this.experimentsRepository = experimentsRepository;
        this.permissionsAwareExperimentRepository = permissionsAwareExperimentRepository;
    }

    public StartExperimentCommand startExperimentCommand(
            String experimentId,
            StartExperimentProperties properties) {
        Preconditions.checkNotNull(experimentId);
        Preconditions.checkNotNull(properties);
        return new StartExperimentCommand(
               experimentsRepository,
               properties,
                permissionsAwareExperimentRepository,
               experimentId
        );
    }
}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/domain/experiments/administration/start/StartExperimentCommandFactory.kt
package pl.allegro.experiments.chi.chiserver.domain.experiments.administration.start

import org.springframework.stereotype.Component
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentsRepository
import pl.allegro.experiments.chi.chiserver.domain.experiments.administration.PermissionsAwareExperimentGetter

@Component
class StartExperimentCommandFactory(val experimentsRepository: ExperimentsRepository,
                                    val permissionsAwareExperimentGetter: PermissionsAwareExperimentGetter) {

    fun startExperimentCommand(
            experimentId: String,
            properties: StartExperimentProperties) : StartExperimentCommand {
        return StartExperimentCommand(
                experimentsRepository,
                properties,
                permissionsAwareExperimentGetter,
                experimentId)
    }
}